Itd AI guidance (#1950)
CodeYourFuture/curriculum: common-content/en/module/itd/step-3/instructions/index.md

Reviewer: This is the longest of the steps they will need to read. Would it be possible to split up into separate pages? There are lots of sections, some of these should be grouped together, but I'm not sure of the best way

1. What is AI
2. How to use it
3. Why we need to be careful
4. The acceptable use policy
5. The CV exercise
6. the questions

Maybe 1,2 & 3,4 & 6,7

Breaking these up a little would be a different layout to the other steps, but might help to get the point across without being a huge wall of text.

Author: Happy to give that a go

Because AI tools "guess" the text to generate, they can make mistakes or make up things that don’t exist (hallucinations). For instance, a famous hallucination from 2024 was that many AI tools said there were two 'r's in the word strawberry.

AI hallucinations can be costly to a programmer. For instance, it has been reported that AI can make up names of code packages which don't exist, and direct a programmer to use the package in their code. This comes with a security risk: malicious packages can get created with names often generated by AI to exploit their "predictions" and get coders to import viruses into their code.

This means that all output from AI is potentially untrustworthy, we should keep a critical eye on the output of AI tools.

> ### Why do we need to be careful as ITD trainees?

Reviewer 2: I'd be tempted to split this into two sub-sections:

1. We need to build understanding
2. Sometimes it lies

I think the "we need to question and verify and explain" could do with being split into those two sections too:
* In the "building understanding" section we can talk more about how we should be able to spot problems in AI output, and we should be able to delete AI output and write it ourselves because we know how.
* In the "sometimes it lies" section, we can talk a bit about looking up other sources

At {{<our-name>}}, we prioritise building understanding so trainees can be employable in the tech industry. Employability in software engineering involves knowing how to debug broken code, find information you might not already know, and break down problems to find a solution. {{<our-name>}} mentors during the ITD and beyond will try to help you without giving away the answers or writing the code for you so you learn how to overcome hurdles yourself and apply that skill in the workplace when mentors aren't around.

**Why does it matter how we learn at {{<our-name>}}? What does it have to do with AI?**

> Well, it can be easy for AI tools to do the hard work for us, e.g. solving the problem, fixing errors, and writing emails for us. But by making sure we do that work ourselves, we build the skills that will help us in our careers. When AI does the work for us, we may think we're progressing when we're actually not building skills that will make us employable.

Reviewer 2: ```suggestion
Well, it can be easy for AI tools to do the hard work for us, e.g. solving the problem, fixing errors, and writing emails for us. But when AI does the work for us, we may think we're making progress and learning when we're not. We're actually not building skills that will make us employable. Instead we need to make sure we do that work ourselves, so that we build the skills that will help us in our careers.
```

AI can make mistakes when it's guessing what output to return to the user, but it can also be led to give incorrect information when a prompt isn't accurate. Which means we need to ask good questions (or prompts). As ITD learners, we are early in the {{<our-name>}} journey, so we are still building our foundational knowledge, which means it is more likely we will ask AI inaccurate questions and AI will give misleading results. It's something to be aware of and cautious about - the accuracy of the questions we ask is important.

For now, remember to question and verify what AI tools give us back. If we can't explain it, or verify the information with our own fact-finding, we shouldn't use it.

### What's acceptable AI use in ITD

Generally, AI might be a good tool for:

- Explaining tech concepts in plain language
- Answering "what does this mean?" questions
- Getting a starting point for research

You need to exercise CAUTION and avoid the following situations:

> - Letting AI write your CV

Reviewer 2: ```suggestion
- Letting AI write something for you (e.g. your CV)
```

And let's explain _why_ this is bad - that it will generate generic CVs which don't sell you, which aren't personal, that use filler words, that make things up about your background, etc.

Maybe we could even include side-by-side a bad AI generated CV and a good personally written CV, and give trainees a prompt for some criteria to use to compare them, and have them submit their comparison? (On average our ITD trainees don't know what a good CV looks like, so I think we would need to give them criteria to evaluate from)

Author: **Thought dump after two weeks of chewing on how to make AI CV creation make sense**

Rn the learning objectives are going in the direction of:
1. Highlight the ways AI can make mistakes or go wrong.
2. Highlight the disadvantages of AI-generated work.
3. Identify good and bad uses of AI for learning during the ITD. 
4. Create a Google Doc with answers to questions about AI usage. 

Point 2 is where CV comparison could come into play. Issues I'm running into atm:

> Upskilling ITD folks to know what a good CV looks like and how to spot issues. A lot of content for a page focused on AI. 

Workaround: I am delegating upskilling trainees in what a good CV looks like by pointing to the guide and evaluator tool. It feels like a lot to explain in the prep. 

> I can currently force AI to hallucinate if I give it a more mid/senior role and a junior "profile" and ask it to generate a CV for it (just getting it to write a CV based on a profile was actually hard to make look bad, common LLMs seem to reasonably make a CV on its own - ChatGPT is fluffier than claude for sure). 
Now the issue is, getting people to spot the hallucinations and fluffed up language is hard when they have no idea about the language, ideas, engineering concepts mentioned. 

Workaround: I can try to make the profile someone who has warehouse/taxi/shop assistant experience and go for a more customer service/intern/office role. **But again hard to tell if people will have enough understanding of the skills to spot the hallucination.** 

If the core q is "why do I bother writing a CV myself" we can drill in these advantages:
- accuracy in a CV, hallucinations avoided
- English lang skills
- Understanding the contents of the CV for interview fluency

And an alternative good use of AI might be finding out more about what a good structure of a CV would be for a uk-based SE, and a bad use would be "write a CV for me" (and why these are good and bad). 

**I'm still chewing on this but needed to put a bookmark in where I'm at with this**

- We're actually not building skills that will make us employable.
- We're not understanding why we do certain things and what happens when we do them incorrectly.

We need to do that work ourselves, so that **we build the deeper skills that will help us in our careers**.

##### How do we use AI to build understanding?

**1. Ask "why", not "how"**

AI is very good at answering "how do I do X?" questions. It is less good at explaining the deeper principles behind how we do things.

Develop the habit of following up with questions that go deeper into understanding, e.g:

- "Why does this approach work?"
- "What are the alternatives?"
- "When would this not work?"

**2. Check if you've learnt something**

There is a meaningful difference between "I used AI to help me understand this concept" and "I used AI to avoid engaging with this concept."

The first makes you stronger. The second leaves a gap that will show itself later, often at the worst moment, like a job interview.

> **Ask yourself: “Will I need to ask AI the same question tomorrow?”**

Reviewer: Is this meant to be instruction 3 following 1 and 2 above?

Author: I have no idea if this still applies to the updates I've made as I've juggled stuff around-  maybe have another look at let me know if I've made this make more sense

Reviewer: On re-reading, I think this part is fine now

Here are two ways to think critically about AI output:

##### 1. Challenge output

We have observed that more experienced developers interact with AI very differently from novices: they push back, ask for alternatives, get justification for the response from AI tooling.

Build this habit early. Don't just defer to AI output without questioning it.

Ask AI for a different approach. Ask it to defend its answer. Ask it what could go wrong.

##### 2. Always Verify

AI makes mistakes. It can produce code that looks correct but contains subtle errors, outdated practices, or security vulnerabilities. It can confidently agree with ideas that are wrong. Your job is always to check against documentation, against your own understanding, by running tests and more.

Look for **verified sources**:

> - Ask AI to cite it's sources, where has it found it's information?

Reviewer: ```suggestion
- Ask AI to cite its sources and then check you trust where it found its information.
```

You need to exercise CAUTION and avoid the following situations:

- Letting AI write something for you (e.g. your CV)
- Copying AI explanations without checking your understanding
- Just accepting the AI answer without exploring further

#### Why is it bad for AI to write for us?

Let's take the example of a CV. We ask you to write your CV and keep it updated. This has the following knock on benefits to your skills:

- You improve your written English skills.
- You are forced to think about your skills history and create your own narrative about your skills.
- By writing it yourself you will remember your CV content more, which helps with interview technique (recall of your skills).
- By writing it yourself you understand the structure and content of the CV, making it easier to make quick edits and updates.

And what happens when AI writes a CV for us:

- You don't build English language instincts, you only read an example of written English. You don't make your own mistakes so you don't learn what is right or wrong. If you get a perfect instance of English generated in the first go, you are less likely to know what mistakes to correct in the future when you write your own words.

> - You don't build a mental map of your CV through structuring it yourself. Harder to recall your own CV in interviews.

Reviewer: ```suggestion
- You don't build a mental map of your CV through structuring it yourself. This makes it harder to recall your own CV in interviews.
```

- Your CV lacks personalised tone of voice. CV comes out generic. Harder to catch the eye of recruiters, easier for recruiters to spot AI generation as it will sound like generated output, easier for the recruiter to discount your CV.
- Big risk of AI making things up about your background - remember AI hallucinates.

#### How to check if you're using AI right

> Ask yourself: **Can you do this task next time without using AI?**

Reviewer: I think this is repeating what was already suggested earlier, but at this stage this is probably a good idea.

### Questions

#### Question One

A {{<our-name>}} trainee is assigned coursework on a {{<our-name>}} course. The assignment is to build a homepage for a library that displays its twenty newest books. It requires that they use CSS Grid (a CSS feature used to display HTML elements in a grid format). They think that ChatGPT could help them complete this task.

Which of the prompts below is an appropriate use of AI by the trainee?

1. "Generate a webpage to display information about books in a grid format."
2. "Use CSS Grid to build a webpage to display twenty digital books."
3. "Build a homepage for a library website."
4. "I want to learn more about CSS Grid. Please explain the core concepts simply with examples."

#### Question Two

A {{<our-name>}} trainee prompts an AI tool to help them understand a complex CSS topic. The information the AI gives doesn't sound right.

Which of the below might be appropriate reactions from the trainee?

1. Googling the topic and looking for reliable sources to clarify the point.

> 2. Trying to persuade the AI to check again.

Reviewer: I'm not sure how ITD would normally handle this. Is this a multiple-choice, or are these suggestions for the trainee to expand on in their answer? Would this be better as an open-ended question?

Author: querying CJ and curriculum channel

3. Accepting this as the truth because the AI seemed to know what it was talking about.
4. Asking a volunteer whether the AI is right.
5. Asking another AI tool. 

#### Question Three

A {{<our-name>}} trainee balances full-time work and family responsibilities with studying at {{<our-name>}}. They are struggling to keep up with the course. In particular, they are struggling to understand some fundamental concepts.

> Which of the below are NOT appropriate actions by the trainee?

Reviewer: I would invert this to focus on the positive actions

1. Asking volunteers for guidance at a session.

> 2. Giving an AI tool their assigned problems and asking it to solve them for them. They then copy and paste the output, making no edits.

Reviewer: ```suggestion
2. Giving an AI tool their assigned problems and asking it to solve them. They then copy and paste the output, making no edits.
```
